File: cpp/2958.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        map<int, int> cnt;
        int ans = 0, cur = 0;
        int l = 0, r = 0;
        int len = nums.size();
        while (r < len) {
            ++cur;
            ++cnt[nums[r]];
            while (cnt[nums[r]] > k) {
                --cur;
                --cnt[nums[l]];
                if (cnt[nums[l]] == 0) {
                    cnt.erase(nums[l]);
                }
                ++l;
            }
            ans = max(ans, cur);
            ++r;
        }
        return ans;
    }
};

File: python3/2958.py
from typing import List
from collections import defaultdict
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        cnt = defaultdict(int)
        ans, cur = 0, 0
        l, r = 0, 0
        n = len(nums)
        while r < n:
            cur += 1
            cnt[nums[r]] += 1
            while cnt[nums[r]] > k:
                cur -= 1
                cnt[nums[l]] -= 1
                if cnt[nums[l]] == 0:
                    cnt.pop(nums[l])
                l += 1
            ans = max(ans, cur)
            r += 1
        return ans